### DISCUSSION TITLE
Why use Makefile?

### DISCUSSION THREAD

**Discussion by nikhenri** [cocotb/cocotb#3177] [Proposals]
Hi,
Let me first say that cocotb is amazing.

There's a thing that I don't understand, it's the reliance on makefiles for operation.
As a Windows user, this is annoying because the makefiles call UNIX commands (`tr`, `tee` etc.) that are absent in a typical Windows installation.

I cannot be sure for all supported simulator, I only use ModelSim and QuestaSim, but in my case what the Makefile does is simple.
It sets some environment variables and create a small file "runsim.do" and finally launch the simulator sourcing the file.
In fact, is so simple that now that I have run it once and saw what it does, I don't use it any more, I just generate the runsim.do with a Python script on my own (not OS dependent).
Knowing that the user has obviously Python installed, I wonder why all the complex set of makefiles is needed just to create a small TCL script to drive tests, when Python could be used instead.
It seems to me that it would be easier and more natural for the same job, and as an added bonus would be OS independent and not require MSYS2 on Windows (important for locked-down workstations).
Instead of using a "root" Makefile to configure paths etc., a INI file or any kind of configuration file could be used.
Finally, if using `make` is so important for some users (because of old habits, muscle memory, etc), a simple Makefile wrapper could be written that would simply relay the job to the Python scripts.
Does that make sense? Is there some serious business performed in the makefiles that I missed that warrant their usage?

**Comment by eric-wieser** [comment]
Makefiles are very good at caching dependencies and not rebuilding things.

Some alternatives that achieve that same goal, in no particular order:
* SCons (python-based)
* Bazel (python-inspired)

Both are much less ubiquitous than Make!

**Reply by Jarhmander** [reply]
Well, of course, that's the purpose of `make` after all.

But it seems here that the job is so trivial and so short (the generated script is not much more than 10 lines IIRC), it could be simpler and safer to just generate the script every time.

Let's turn the question a bit differently, then. What would be a major drawback, as a user, to invoke `make -B` instead of just `make`, except for the longer typing? Note that I haven't tried it, and I know for a fact that some Makefile constructions e.g. ["match-anything rules"](https://www.gnu.org/software/make/manual/html_node/Match_002dAnything-Rules.html) tends to break things when using the `-B` flag.

What are the complex dependencies to track and what needs not to be redone every time?

**Comment by imphil** [comment]
What you're describing (a replacement for the cocotb-shipped Makefiles) already exists in various forms. There's cocotb-test, and there's the cocotb runner (experimental in master currently), and there is experimentation with edalize and cocotb. However, the Makefile approach remains the oldest, and also most tested approach. 

There's nothing in cocotb that requires the use of Makefiles, it's just the "default" run system that we provide. As you say, it's easy to just call your simulator directly and have it load cocotb. That's documented at https://docs.cocotb.org/en/stable/custom_flows.html

**Comment by nikhenri** [comment]
I understand now that the makefile is not necessary, the installation page show it as it must be part of the installation on Windows
<img width="547" alt="image" src="https://user-images.githubusercontent.com/61157967/212160103-36f0a8cf-e014-4419-baab-acc98778a191.png">

Then in the quickstart is also show as the make is necessary to run the example
<img width="584" alt="image" src="https://user-images.githubusercontent.com/61157967/212160430-ce1cd2aa-d295-435f-a5f7-a485911a02c5.png">

In my opinion it would be nice if we could run the getting started without make, something like
Run 'python run.py' instead of 'make'

**Comment by imphil** [comment]
The default way of using cocotb is using a Makefile. The installation instructions tell you how to get a system that can deal with that. 

There are efforts to use something other than Makefiles as default (e.g., cocotb runner). Are you interested in joining the effort to make this way of running cocotb the default way? Note that this requires a fair amount of experimentation with different tools, and user support -- the tricky part is not the build system, but the various quirks that need to be applied and updated to get the different tools working.
